parsl/data_provider/zip.py

```diff
@@ -0,0 +1,84 @@
+import filelock
+import logging
+import os
+import parsl
+import zipfile
+
+from typing import Tuple
+
+from parsl.data_provider.staging import Staging
+from parsl.data_provider.files import File
+
+
+logger = logging.getLogger(__name__)
+
+
+class ZipFileStaging(Staging):
+    """A stage-out provider for zip files.
+
+    This provider will stage out files by writing them into the specified zip
+    file.
+
+    The filename of both the zip file and the file contained in that zip are
+    specified using a zip: URL, like this:
+
+    zip:/tmp/foo/this.zip/inside/here.txt
+
+    This URL names a zip file ``/tmp/foo/this.zip`` containing a file
+    ``inside/here.txt``.
+
+    The provider will use the Python filelock package to lock the zip file so
+    that it does not conflict with other instances of itself. This lock will
+    not protect against other modifications to the zip file.
+    """
+
+    def can_stage_out(self, file: File) -> bool:
+        logger.debug("archive provider checking File {}".format(repr(file)))
+        return file.scheme == 'zip'
+
+    def stage_out(self, dm, executor, file, parent_fut):
+        assert file.scheme == 'zip'
+
+        zip_path, inside_path = zip_path_split(file.path)
+
+        working_dir = dm.dfk.executors[executor].working_dir
+
+        if working_dir:
+            file.local_path = os.path.join(working_dir, inside_path)
+
+            # TODO: I think its the right behaviour that a staging out provider should create the directory structure
+            # for the file to be placed in?
+            os.makedirs(os.path.dirname(file.local_path), exist_ok=True)
+        else:
+            raise RuntimeError("zip file staging requires a working_dir to be specified")
+
+        stage_out_app = _zip_stage_out_app(dm)
+        app_fut = stage_out_app(zip_path, inside_path, working_dir, inputs=[file], _parsl_staging_inhibit=True, parent_fut=parent_fut)
+        return app_fut
+
+
+def _zip_stage_out(zip_file, inside_path, working_dir, parent_fut=None, inputs=[], _parsl_staging_inhibit=True):
+    file = inputs[0]
+
+    os.makedirs(os.path.dirname(zip_file), exist_ok=True)
+
+    with filelock.FileLock(zip_file + ".lock"):
+        with zipfile.ZipFile(zip_file, mode='a', compression=zipfile.ZIP_DEFLATED) as z:
+            z.write(file, arcname=inside_path)
```

Comment:
Is there any value in attempting to test the lock?  It's a new requirement in `requirements.txt` and as discussed out-of-band, a real consideration, so perhaps worth an AT of some kind?